src/css3_parser/parser.rs

```diff
@@ -0,0 +1,500 @@
+        }
+
+        list
+    }
+
+    /// ```txt
+    /// CSSStyleDeclaration
+    ///     : PropName ':'  PropValue ';'
+    ///     ;
+    /// ```
+    fn css_style_declaration(&mut self) -> CSSStyleDeclaration {
+        let name = self.prop_name();
+        self.consume(TokenType::Colon);
+        let value = self.prop_value();
+        self.consume(TokenType::SimiColon);
+
+        CSSStyleDeclaration { name, value }
+    }
+
+    /// ```txt
+    /// PropName
+    ///     : String
+    ///     ;
+    /// ```
+    fn prop_name(&mut self) -> String {
+        let token = self.consume(TokenType::Identifier);
+        token.value.to_string()
+    }
+
+    /// ```txt
+    /// PropValue
+    ///     : String
+    ///     ;
+    /// ```
+    fn prop_value(&mut self) -> String {
+        let token = self.consume(TokenType::Identifier);
+        token.value.to_string()
+    }
+
+    fn consume(&mut self, token_type: TokenType) -> Token {
+        if let Some(token) = self.lookahead.clone() {
+            if token.token_type != token_type {
+                panic!(
+                    "Unexpected token: '{:?}', expected: '{:?}'. Got '{}' at '{}'",
+                    token.token_type, token_type, token.value, self.tokenizer.cursor
+                )
+            }
+
+            // Advance to the next token
+            self.lookahead = self.tokenizer.get_next_token();
+
+            println!("next token: {:#?}", self.lookahead);
+            return token.clone();
+        }
+
+        panic!("Unexpected end of input, expected: {:?}", token_type)
+    }
+
+    fn is_next_token(&self, token_type: TokenType) -> bool {
+        if let Some(token) = self.lookahead.clone() {
+            return token.token_type == token_type;
+        }
+
+        false
+    }
+
+    fn is_next_tokens(&self, token_types: Vec<TokenType>) -> bool {
+        for token_type in token_types {
+            if self.is_next_token(token_type) {
+                return true;
+            }
+        }
+        false
+    }
+
+    fn get_next_token_type(&self) -> Option<TokenType> {
+        if let Some(token) = self.lookahead.clone() {
+            return Some(token.token_type);
+        }
+
+        None
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn should_parse_css() {
+        let mut parser = CSS3Parser::new();
+        let style_sheet = parser.parse(
+            r#"
+            /* Shoud skip comments */
+
+            .header {
+                width: 10px;
+                font-size: 1rem;
+            } 
+
+            #nav {
+                height: 200px;
+            }
+
+            @media screen, printer, all and (max-width: 600px) {
+                .header {
+                    display: flex;
+                }
+
+                body {
+                    max-height: 100vh;
+                }
+            }
+
+
+            @media all {}
+        "#,
+        );
+
+        assert_eq!(
```

Comment:
I love this test.  It would be nice if we had things up so that the testing of the HTML parsing was like this as well.